### PR TITLE
$p->resolve($p) rejects with an error message, and only one call to resolve considered

### PR DESCRIPTION
**Pull request by akarelas**
### Summary
Sort of solves bug #1517 (`$p->resolve($p)` should reject an exception). I say "sort of" because the reason of the rejection is not a TypeError object (itself a subclass of Error), but a simple text message, because I don't know how to produce errors (and if they are at all relevant to Perl). Nevertheless, with this PR I provide a solution the maintainers can modify to make an Error if they so wish.

**Edit:** This PR now also solves bug #1515.

### Motivation
This change makes Mojo::Promise objects more in accordance with the spec (https://tc39.es/ecma262/#sec-promise-resolve-functions)

### References
* Bug #1517 
* https://tc39.es/ecma262/#sec-promise-resolve-functions point 7
* https://tc39.es/ecma262/#sec-promise-resolve-functions points 5 & 6

### CODE REVIEW

File: lib/Mojo/Promise.pm
Comment (jhthorsen) [inline lib/Mojo/Promise.pm:219]: I don't think this is run through perltidy. I suggest this formatting:

```
    (refaddr $results[0] == refaddr $self)
      ? $self->reject('Chaining cycle detected.')
      : $results[0]->then(sub { $self->resolve(@_); () }, sub { $self->reject(@_); () });
```

**Review by jhthorsen — CHANGES_REQUESTED** [review]
This PR cannot be accepted without any accommodating tests.

I'm unsure if this is a real issue though or simply an academic one, but I'm positive to the change.

File: lib/Mojo/Promise.pm
Comment (akarelas) [reply lib/Mojo/Promise.pm]: I don't have access to perltidy right now. You'd have to run it through perltidy yourself, if that's not a huge problem for you. Thanks.

**Comment by christopherraa** [comment]
I'm also positive to the change, but there needs to be tests. Please add tests for this change and I'll follow up by reviewing it.

**Comment by akarelas** [comment]
This PR now solves two issues (#1515 and #1517), contains tests, and have passed everything (`lib/Mojo/Promise.pm` and `t/mojo/promise.t`) through perltidy.

File: lib/Mojo/Promise.pm
Comment (kraih) [inline lib/Mojo/Promise.pm:217]: We don'y use underscore names.

Comment (kraih) [reply lib/Mojo/Promise.pm:217]: Or this `... or return ...` style.

File: t/mojo/promise.t
Comment (kraih) [inline t/mojo/promise.t:521]: Other tests don't use `@results` for both.

File: t/mojo/promise.t
Comment (kraih) [inline t/mojo/promise.t:524]: Inconsistent description.

**Review by kraih — DISMISSED** [review]

**Comment by kraih** [comment]
Your commit messages are really bad, to the point where i will have to reject the PR for it. And you did not squash the commits.

File: lib/Mojo/Promise.pm
Comment (kraih) [inline lib/Mojo/Promise.pm:222]: Isn't the JavaScript exception `Chaining cycle detected for promise`?

File: lib/Mojo/Promise.pm
Comment (akarelas) [reply lib/Mojo/Promise.pm]: addressed

File: lib/Mojo/Promise.pm
Comment (akarelas) [reply lib/Mojo/Promise.pm]: Addressed.

File: t/mojo/promise.t
Comment (akarelas) [reply t/mojo/promise.t]: Addressed.

File: t/mojo/promise.t
Comment (akarelas) [reply t/mojo/promise.t]: Addressed.

File: lib/Mojo/Promise.pm
Comment (akarelas) [reply lib/Mojo/Promise.pm]: Addressed.

**node.js** exception is `Chaining cycle detected for promise [Promise]`. Remembering how **node.js** outputs `[object Object]` when you stringify an object (as in `console.log('hi: ' + {})`), it came to my mind that it would be more correct, and helpful to developers who are investigating their own bugs at the same time, if I stringified $self (a promise) in this message too (as in: `"Chaining cycle detected for promise $self"`).

If you don't agree with this I will remove $self from that string.

File: lib/Mojo/Promise.pm
Comment (kraih) [inline lib/Mojo/Promise.pm:222]: What is the use of including `$self` there?

**Comment by akarelas** [comment]
> Your commit messages are really bad, to the point where i will have to reject the PR for it. And you did not squash the commits.

Fixed messages and squashed commits, but now the message takes 2 lines.

File: t/mojo/promise.t
Comment (kraih) [inline t/mojo/promise.t:525]: How is that better than `is_deeply \@errors, [], ...`?

File: lib/Mojo/Promise.pm
Comment (akarelas) [reply lib/Mojo/Promise.pm:222]: The use is that this kind of cycling error is only going to happen in a bug. So I included `$self` here to allow the developer to debug and figure out which of their promises took part in the cycle. (I'm assuming the developer can insert debugging code in their own code to display the memory address of their promises, but shouldn't have to tweak the Mojolicious library's code to display the address of the promise that took part in the cycle).

Comment (akarelas) [reply lib/Mojo/Promise.pm:222]: Plus, node.js also displays `...for promise [Promise]` in their error message, which kind of reminds how objects are stringified (`[object Object]`), so perhaps including `$self` increases the consistency with node.js.

File: t/mojo/promise.t
Comment (kraih) [inline t/mojo/promise.t:529]: Is it resolve "to" or resolve "with" unsettled promise?

Comment (akarelas) [reply t/mojo/promise.t:529]: I do not know. Couldn't find either of the two in past tests. I believe I haven't used both verbs in my commit.

But I believe "resolve to" conveys the message in a more clear way in this case, because "Resolve with unsettled promise" could also mean "calling `resolve` on an unsettled promise" which is not what I meant.

* This promise resolves to 5
* This promise resolves with 5

Maybe both expressions are valid, but I can't say. I'm not that good with English prepositions.

**Comment by akarelas** [comment]
I created this looping test script, which tests 10 settling cases for Promises. e.g. `Mojo::Promise->resolve($p);` and `$p->reject($p)`, etc.

Maybe you might want to keep a copy to test all 10 cases next time you change something in Mojo::Promise. I created it to test for bugs, and to check whether I fixed them in my PR.

https://gist.github.com/akarelas/1cddeda23e6a408c31929b3a1993964f

File: t/mojo/promise.t
Comment (akarelas) [reply t/mojo/promise.t]: Since I saw you were already using the preposition "with" in similar -although non-ambiguous- situations inside `mojo/promise.t`, I decided to keep the same preposition, and addressed the ambiguity by renaming "Resove to unsettled promise" to "Settle with unsettled promise" and added the `$p->reject($p2)` case inside that subtest, too. Also I renamed "Resolve to self" to "Resolve with self", as that was already non-ambiguous.

To make sure the case "Reject with self" is covered, I renamed the subtest "Resolve to self" to "Settle with self" and added the "Reject with self" case in there too.

**Comment by shadowcat-mst** [comment]
If you mean to resolve a promise that itself is unsettled,

  $unsettled->resolve(...)

would be "resolve unsettled promise"

If you mean to resolve a promise by passing another unsettled promise,

  $promise->resolve($unsettled)

would be "resolve with unsettled promise"

File: t/mojo/promise.t
Comment (akarelas) [reply t/mojo/promise.t]: shadowcat-mst below provided the answer to your question: "resolve "to" or resolve "with"" (it is "resolve with"). I have thus renamed the subtests, and included more cases in them (with "reject" instead of "resolve").

Anything else you need to accept this PR?

**Comment by akarelas** [comment]
@christopherraa , @jhthorsen I have addressed all the issues you and kraih had with this bugfix.

Could you please look at it again and approve it, if it passes your criteria?

Thanks.

**Comment by kraih** [comment]
I don't see this PR going anywhere. The complexity is just way too high for the (still questionable) value it would add.

**Comment by akarelas** [comment]
> I don't see this PR going anywhere. The complexity is just way too high for the (still questionable) value it would add.

I'd like to remind that it also solves bug #1515. Anyway, thought it would be cool to solve two bugs with 15 lines of code. Your choice.

File: t/mojo/promise.t
Comment (kraih) [inline t/mojo/promise.t:410]: This block of tests seems completely unrelated to the PR description.

File: t/mojo/promise.t
Comment (kraih) [inline t/mojo/promise.t:352]: These tests are unrelated to your PR, why would you randomly change their description?

**Review by kraih — CHANGES_REQUESTED** [review]
I'll block this PR because it contains random unexplained changes.

**Comment by kraih** [comment]
Has this PR been abandoned?

**Comment by jhthorsen** [comment]
I cannot remember seeing this problem while writing code, so I'm no longer positive to the change. I consider the PR to have academic value, but not something that is frequent enough in real life to be considered.

File: t/mojo/promise.t
Comment (akarelas) [reply t/mojo/promise.t:410]: Indeed, this block was related to the second issue this PR solves, which was added later. (For history reasons I thought I shouldn't have modified any of the past comments, including the title & description). Now the PR's description has been updated to address that.

File: t/mojo/promise.t
Comment (akarelas) [reply t/mojo/promise.t:352]: I was under the impression that a patch needs to leave the repo in a consistent state. It would seem strange to have a subtest that would test "Settle with promise", and another one that would test "Settle with unsettled promise". Because, what would the second subtest's reason for existence be, if its case is already being tested by the first subtest? So I did this modification.

Shall I revert that line?

**Comment by akarelas** [comment]
> Has this PR been abandoned?

No, sorry, I was just busy celebrating Christmas. Now I'm back to working.

**Comment by akarelas** [comment]
> I cannot remember seeing this problem while writing code, so I'm no longer positive to the change. I consider the PR to have academic value, but not something that is frequent enough in real life to be considered.

Indeed, the only reason I made this PR was to improve compliance with the ES spec. None of my code would encounter these issues either, unless my code had a bug. I'm okay with you not merging this PR.

File: t/mojo/promise.t
Comment (kraih) [reply t/mojo/promise.t:410]: PRs trying to resolve multiple issues at once pretty much never get accepted.

**Comment by kraih** [comment]
It seems nobody has ever actually run into these issues and the code complexity is fairly high. So i'm 👎 on this PR.

**Review by jberger — CHANGES_REQUESTED** [review]
After letting this sit this long, I've got to agree. It is a lot of complexity on top of an already complex (of necessity) codebase. If it isn't addressing a real world need (like warning on unhandled errors) i vote no

**Comment by kraih** [comment]
Thanks for your contribution, but i don't see this PR pass a vote.

**Pull request closed without merging**
